### PR TITLE
Translator version 2

### PR DESCRIPTION
**Pull request by AlexandreF-1qbit**
Translator module V2 for converting circuit to and from many formats. 

Main changes:
- Added a `translate_circuit` function (with source and target arguments).
- Added a deprecated decorator.
- Removed warning filters.
- Added translate function from Braket and Json IonQ.

### CODE REVIEW

**Comment by AlexandreF-1qbit** [comment]
If you have any suggestion on the deprecated messages, I am open for discussion!

File: tangelo/linq/translator/translate_braket.py [hunk 1]
```diff
@@ -99,3 +116,42 @@ def translate_braket(source_circuit):
...
+            name = name.rstrip("SHIFT")
+
+        if name in {"H", "X", "Y", "Z", "S", "T", "SWAP"}:
+            gates += [Gate(name, qubits)]
+        elif name in {"RX", "RY", "RZ", "PHASE"}:
+            gates += [Gate(name, qubits, parameter=instruction.operator.angle)]
+        elif name in {"CNOT", "CX", "CY", "CZ", "CSWAP"}:
+            gates += [Gate(name, target=qubits[1:], control=qubits[0])]
+        elif name in {"XX"}:
+            gates += [Gate(name, target=qubits, parameter=instruction.operator.angle)]
```
Comment (JamesB-1qbit) [inline tangelo/linq/translator/translate_braket.py:151]: "XX" is the same as "RX", "RY", "RZ", "PHASE" and can be combined

**Review by JamesB-1qbit — APPROVED** [review]

File: tangelo/linq/translator/translate_braket.py (hunk 1 above)
Comment (AlexandreF-1qbit) [reply tangelo/linq/translator/translate_braket.py]: Indeed!

File: tangelo/linq/translator/translate_braket.py [hunk 2]
```diff
@@ -99,3 +116,40 @@ def translate_braket(source_circuit):
         else:
             raise ValueError(f"Gate '{gate.name}' not supported on backend braket")
     return target_circuit
+
+
+def translate_c_from_braket(source_circuit) -> Circuit:
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_braket.py:121]: I'd rather you do not document type that half-ass it. You haven't done it for any other function, and you're not doing it for the argument on the same line.

I think however we should have a conversation about types in Python as a group

File: tangelo/linq/translator/translate_circuit.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    source = source.lower()
+    target = target.lower()
+
+    if output_options is None:
+        output_options = dict()
+
+    if source == target:
+        return circuit
+    if source != "tangelo":
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_circuit.py:69]: skip a line and add a comment:

`# Convert to Tangelo format if necessary`

File: tangelo/linq/translator/translate_circuit.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        if source not in TO_TANGELO:
+            raise NotImplementedError(f"Circuit conversion from {source} to {target} is not supported.")
+        circuit = TO_TANGELO[source](circuit)
+    if target != "tangelo":
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_circuit.py:73]: skip a line and add a comment:

`# Convert to another target format if necessary`

File: tangelo/linq/translator/translate_circuit.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+def translate_circuit(circuit, target, source="tangelo", output_options=None):
+    """Function to convert a quantum circuit defined within the "source" format
+    to a "target" format.
+
+    Args:
+        circuit (source format): Self-explanatory.
+        target (string): Identifier for the target format.
+        source (string): Identifier for the source format.
+        output_options (dict): Backend specific options (e.g. a noise model).
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_circuit.py:55]: We need to open that up a bit. Pretty sure some backends can specify a number of qubits etc.

File: tangelo/linq/translator/translate_cirq.py [hunk 6]
```diff
@@ -57,7 +60,23 @@ def get_cirq_gates():
     return GATE_CIRQ
 
 
-def translate_cirq(source_circuit, noise_model=None):
+@deprecated("Please use the translate_circuit function.")
+def translate_cirq(source_circuit):
+    """Take in an abstract circuit, return an equivalent cirq QuantumCircuit
+    instance.
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_cirq.py:66]: instance => object

File: tangelo/linq/translator/translate_json_ionq.py [hunk 7]
```diff
@@ -74,3 +92,45 @@ def translate_json_ionq(source_circuit):
...
+        Circuit: the corresponding quantum Circuit in Tangelo format.
+    """
+
+    starting_circuit = Circuit(n_qubits=source_circuit["qubits"])
+
+    gates = []
+    for gate in source_circuit["circuit"]:
+        name = gate["gate"].upper()
+
+        target_qubits = gate.get("target", gate.get("targets"))
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_json_ionq.py:116]: Just wanted to make sure we understand and are ok with what happens when "targets" or "controls" are not present and we call `gate.get("targets")`. It seems that you are handling `None` values afterwards, which may be the defaults.

As long as you're clear with those, feel free to ignore this comment and resolve the situation.

File: tangelo/linq/translator/translate_json_ionq.py [hunk 8]
```diff
@@ -74,3 +92,45 @@ def translate_json_ionq(source_circuit):
...
+
+    Args:
+        source_circuit (dict): representation of the quantum circuit following
+            the IonQ JSON format.
+
+    Returns:
+        Circuit: the corresponding quantum Circuit in Tangelo format.
+    """
+
+    starting_circuit = Circuit(n_qubits=source_circuit["qubits"])
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_json_ionq.py:110]: `initial_state_circuit` or `zero_state_circuit` seem better.
You don't seem to use it until L135, so this intermediate variable may not be useful at all, maybe the expression can directly be injected into L135

File: tangelo/linq/translator/translate_openqasm.py [hunk 9]
```diff
@@ -57,12 +60,47 @@ def translate_openqasm(source_circuit):
...
+        str: the corresponding OpenQASM program, as per IBM Qiskit.
+    """
+    from .translate_qiskit import translate_c_to_qiskit
 
-    return translate_qiskit(source_circuit).qasm()
+    return translate_c_to_qiskit(source_circuit).qasm()
 
 
-def _translate_openqasm2abs(openqasm_str):
+def translate_c_from_openqasm(openqasm_str):
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_openqasm.py:103]: Open question: this explicit conversion only supported a subset of OpenQASM, and was written early 2020.

Can we simply implement this using the reverse qiskit translation combined with https://qiskit.org/documentation/stubs/qiskit.circuit.QuantumCircuit.from_qasm_str.html#qiskit.circuit.QuantumCircuit.from_qasm_str ?

If this is tricky, you can leave this alone for this PR. I suspect some of our circuit data for tests was generated in a particular quasi format, which means things may break. In that case, a follow-up PR can do that.

File: tangelo/linq/translator/translate_qdk.py [hunk 10]
```diff
@@ -46,7 +48,24 @@ def get_qdk_gates():
...
-def translate_qsharp(source_circuit, operation="MyQsharpOperation"):
+@deprecated("Please use the translate_circuit function.")
+def translate_qsharp(source_circuit):
+    """Take in an abstract circuit, generate the corresponding Q# operation
+    (state prep + measurement) string, in the appropriate Q# template. The Q#
+    output can be written to file and will be compiled at runtime.
+
+    Args:
+        source_circuit: quantum circuit in the abstract format.
+        operation (str), optional: name of the Q# operation.
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_qdk.py:59]: im not sure what this argument is, it does not appear in the signature

File: tangelo/linq/translator/translate_qulacs.py [hunk 11]
```diff
@@ -58,7 +60,25 @@ def get_qulacs_gates():
...
+@deprecated("Please use the translate_circuit function.")
+def translate_qulacs(source_circuit):
+    """Take in an abstract circuit, return an equivalent qulacs QuantumCircuit
+    instance. If provided with a noise model, will add noisy gates at
+    translation. Not very useful to look at, as qulacs does not provide much
+    information about the noisy gates added when printing the "noisy circuit".
+
+    Args:
+        source_circuit: quantum circuit in the abstract format.
+        noise_model: A NoiseModel object from this package, located in the
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/translator/translate_qulacs.py:72]: No such argument is referenced

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
A few things. I suggest:

- you import the translate function directly from the translator module.
- we make the translate function visible from `tangelo.linq`

File: tangelo/linq/translator/translate_json_ionq.py (hunk 7 above)
Comment (AlexandreF-1qbit) [reply tangelo/linq/translator/translate_json_ionq.py:116]: When "targets" is not present, there is a "target" (unless there is not circuit, therefore no json dict). When there is no "controls" and no "control", is is handled below.

File: tangelo/linq/translator/translate_braket.py (hunk 2 above)
Comment (AlexandreF-1qbit) [reply tangelo/linq/translator/translate_braket.py]: This was written from the already reviewed function in `translate_qiskit`. Just to be consistent, I removed that ("-> Circuit") everywhere in the translator module.

File: tangelo/helpers/utils.py [hunk 12]
```diff
@@ -40,6 +42,28 @@ def is_package_installed(package_name):
...
+    It results in a warning being emitted when the function is used.
+
+    Ref: https://stackoverflow.com/a/30253848
+
+    Args:
+        custom_message (str): Message to append to the deprecation warning.
+    """
+    def deprecated_wrapper(func):
+        @functools.wraps(func)
+        def new_func(*args, **kwargs):
```
Comment (ValentinS4t1qbit) [inline tangelo/helpers/utils.py:56]: call this "wrapper", that happens to be also its purpose.

**Review by ValentinS4t1qbit — APPROVED** [review]
Good work, thank you for taking our translator to the next level. This will be helpful for any user who wants to jump in and out of Tangelo and play with different frameworks

**Pull request merged**